// [START declaration]
syntax = "proto3";
package message;

import "google/protobuf/timestamp.proto";
// [END declaration]

// [START messages]
message JoinMe {
  string nodeID = 1;
  string IP = 2;
  string port = 3;
}

message Join {
  string nodeID = 1;
  string IP = 2;
  string port = 3;
  int32 row_index = 4;
}

message Node {
  string nodeID = 1;
  string IP = 2;
  string port = 3;
}

message List {
  repeated Node node = 1;
}

message RoutingUpdate {
  bool buddy = 1;
  bool terminal = 2;
  List leaf = 3;
  List neighbours = 4;
  message RoutingEntry {
    int32 index = 1;
    List nodeList = 2;
  }
  repeated RoutingEntry routingEntires = 5;
  int32 prefix_match = 6;
} 

message AllStateUpdate {
  List leaf = 1; 
  repeated List routingTable = 2;
  List neighbours = 3;
}

message GetVal {
  string key = 1;
  Node node = 2;
  string actual_key = 3;
}

message GetValResponse {
  string key = 1;
  string value = 2;
  string actual_key = 3;
}


message SetVal {
  string key = 1;
  string val = 2;
  Node node = 3;
  bool terminal = 4;
}

message Message {
  string type = 1;
  Node sender = 2;
  oneof msg {
    JoinMe joinMeMsg = 3;
    Join joinMsg = 4;
    GetVal getValMsg = 5;
    SetVal setValMsg = 6;
    RoutingUpdate routingUpdate = 7;
    AllStateUpdate allStateUpdate = 8;
    GetValResponse getValResponse = 9;
    DeleteNode deleteNode = 10;
    AddToHashTable addToHashTable = 11;
    ResponseLeafSet responseLeafSet = 12;
    ResponseNeighbourSet responseNeighbourSet = 13;
    RequestRoutingEntry requestRoutingEntry = 14;
    ResponseRoutingEntry responseRoutingEntry = 15;
  }
}

message Response {
  string status = 1;
}

message DeleteNode{
  Node node = 1;
}

message AddToHashTable{
  map<string, string> hashMap = 1;
}

message ShutDown{
  
}

message ResponseLeafSet{
  List leaf = 1;
}

message ResponseNeighbourSet{
  List neighbours = 1;
}

message ResponseRoutingEntry{
  List routingEntry = 1;
}

message RequestRoutingEntry{
  int32 index = 1;
}

// [END messages]